File: algorithms/[234]回文链表/solution.ts
// topics = ["链表", "双指针"]

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function isPalindrome(head: ListNode | null): boolean {
  if (!head || !head.next) return true;

  /**
   * 快慢指针 + 原地翻转链表，使得空间复杂度为常量
   */
  let fast: ListNode | null = head;
  let slow: ListNode | null = head;

  while (fast && fast.next) {
    fast = fast.next.next;
    slow = slow!.next;
  }

  let prev: ListNode | null = null;
  while (slow) {
    const next = slow.next!;
    slow.next = prev;
    prev = slow;
    slow = next;
  }

  while (prev) {
    if (prev.val !== head!.val) {
      return false;
    }
    prev = prev!.next;
    head = head!.next;
  }

  return true;
}

File: algorithms/[234]回文链表/solution.py
# topics = ["链表", "双指针"]

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

from __future__ import annotations
from typing import Union


class ListNode:
    def __init__(self, x: int = 0):
        self.val = x
        self.next: Union[ListNode, None] = None


class Solution:
    def isPalindrome(self, head: ListNode) -> bool:
        fast = slow = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

        prev = None
        while slow:
            next = slow.next
            slow.next = prev
            prev = slow
            slow = next

        while prev:
            if prev.val != head.val:
                return False
            prev = prev.next
            head = head.next
        return True
